pkg/kubectl/cmd/plugin.go

```diff
@@ -61,36 +65,122 @@ func NewCmdPlugin(f cmdutil.Factory, in io.Reader, out, err io.Writer) *cobra.Co
+func NewCmdForPlugin(f cmdutil.Factory, plugin *plugins.Plugin, runner plugins.PluginRunner, in io.Reader, out, errout io.Writer) *cobra.Command {
 	if !plugin.IsValid() {
 		return nil
 	}
 
-	return &cobra.Command{
+	cmd := &cobra.Command{
 		Use:     plugin.Name,
 		Short:   plugin.ShortDesc,
 		Long:    templates.LongDesc(plugin.LongDesc),
 		Example: templates.Examples(plugin.Example),
 		Run: func(cmd *cobra.Command, args []string) {
-			ctx := plugins.RunningContext{
-				In:         in,
-				Out:        out,
-				ErrOut:     errout,
-				Args:       args,
-				Env:        os.Environ(),
-				WorkingDir: plugin.Dir,
+			if len(plugin.Command) == 0 {
+				cmdutil.DefaultSubCommandRun(errout)(cmd, args)
+				return
+			}
+
+			cfg, err := f.ClientConfig()
+			cmdutil.CheckErr(err)
+
+			runningEnvProvider := &plugins.MultiRunningEnvProvider{
+				&plugins.PluginCallerEnvProvider{},
+				&plugins.OSEnvProvider{},
+				&plugins.PluginDescriptorEnvProvider{
+					Plugin: plugin,
+				},
+				&flagsPluginEnvProvider{
+					cmd: cmd,
+				},
+				&factoryAttrsPluginEnvProvider{
+					factory: f,
+				},
+				&restClientConfigPluginEnvProvider{
+					cfg: cfg,
+				},
+			}
+
+			runningContext := plugins.RunningContext{
+				In:          in,
+				Out:         out,
+				ErrOut:      errout,
+				Args:        args,
+				EnvProvider: runningEnvProvider,
+				WorkingDir:  plugin.Dir,
 			}
-			if err := runner.Run(plugin, ctx); err != nil {
+
+			if err := runner.Run(plugin, runningContext); err != nil {
 				cmdutil.CheckErr(err)
 			}
 		},
 	}
+
+	for _, childPlugin := range plugin.Tree {
+		cmd.AddCommand(NewCmdForPlugin(f, childPlugin, runner, in, out, errout))
+	}
+
+	return cmd
+}
+
+type flagsPluginEnvProvider struct {
+	cmd *cobra.Command
+}
+
+func (p *flagsPluginEnvProvider) Env() ([]string, error) {
+	prefix := "KUBECTL_PLUGINS_GLOBAL_FLAG_"
+	env := []string{}
+	p.cmd.Flags().VisitAll(func(flag *pflag.Flag) {
+		env = append(env, plugins.FlagToEnv(flag, prefix))
+	})
+	return env, nil
+}
+
+type factoryAttrsPluginEnvProvider struct {
+	factory cmdutil.Factory
+}
+
+func (p *factoryAttrsPluginEnvProvider) Env() ([]string, error) {
+	cmdNamespace, _, err := p.factory.DefaultNamespace()
+	if err != nil {
+		return []string{}, err
+	}
+	return []string{fmt.Sprintf("%s=%s", "KUBECTL_PLUGINS_CURRENT_NAMESPACE", cmdNamespace)}, nil
+}
+
+type restClientConfigPluginEnvProvider struct {
+	cfg *restclient.Config
+}
+
+func (p *restClientConfigPluginEnvProvider) Env() ([]string, error) {
+	prefix := "KUBECTL_PLUGINS_REST_CLIENT_CONFIG_"
+	env := []string{}
+	env = append(env, plugins.FieldToEnv("Host", p.cfg.Host, prefix))
+	env = append(env, plugins.FieldToEnv("APIPath", p.cfg.APIPath, prefix))
+	env = append(env, plugins.FieldToEnv("Prefix", p.cfg.Prefix, prefix))
+	env = append(env, plugins.FieldToEnv("Username", p.cfg.Username, prefix))
+	env = append(env, plugins.FieldToEnv("Password", p.cfg.Password, prefix))
```

Comment:
I'd like to provide plugin authors a way to access the api (other that invoking `kubectl` through the `KUBECTL_PLUGINS_CALLER` env we are introducing here), so both ideas work for me. I agree with the concerns of these sensitive data passing around and the proxy is definitely easier for plugin authors, but do you think it would be easier to get approved? If so, that was part of my original plugins R&D and should not be too hard to adjust/improve be added here or in a follow-up PR.